src/backend/access/transam/twophase.c

```diff
@@ -1883,7 +1917,94 @@ GetOldestPreparedTransaction()
 void
 StandbyRecoverPreparedTransactions(bool overwriteOK)
 {
-	elog(ERROR, "Hot Standby not supported");
+	prpt_map   *entry;
+	XLogRecPtr	xlogrecptr;
+	XLogReaderState *xlogreader;
+	XLogRecord *record;
+	char	   *errormsg;
+	HASH_SEQ_STATUS hsStatus;
+
+	/* Limit the usage only for TAP tests for now. */
+	if (Gp_role != GP_ROLE_UTILITY)
+	{
+		elog(ERROR, "Hot Standby is not supported");
+	}
+
+	if (crashRecoverPostCheckpointPreparedTransactions_map_ht == NULL)
+		return;					/* No prepared transactions to recover */
+
+	/* Allocate XLog reader once for all transactions */
+	xlogreader = XLogReaderAllocate(&read_local_xlog_page, NULL);
+	if (!xlogreader)
+		ereport(
+				ERROR,
+				(errcode(ERRCODE_OUT_OF_MEMORY), errmsg("out of memory"),
+		   errdetail("Failed while allocating an XLog reading processor.")));
+
+	/* Iterate through all entries in the hash table */
+	hash_seq_init(&hsStatus,
+				  crashRecoverPostCheckpointPreparedTransactions_map_ht);
+
+	while ((entry = (prpt_map *) hash_seq_search(&hsStatus)) != NULL)
+	{
+		TransactionId xid = entry->xid;
+		TwoPhaseFileHeader *hdr;
+		TransactionId *subxids;
+		char	   *buf;
+		char	   *bufptr;
+		int			i;
+
+		xlogrecptr = entry->xlogrecptr;
+
+		if (xlogrecptr == InvalidXLogRecPtr)
+			continue;
+
+		record = XLogReadRecord(xlogreader, xlogrecptr, &errormsg);
+		if (record == NULL)
+		{
+			if (errormsg != NULL)
+			{
+				elog(
+					 WARNING,
+				 "could not read WAL record for prepared transaction %u: %s",
+					 xid, errormsg);
+			}
+			continue;			/* Skip this transaction */
+		}
+
+		buf = XLogRecGetData(record);
+		hdr = (TwoPhaseFileHeader *) buf;
+
+		/* Skip if this transaction has been committed/aborted already */
```

Comment:
Should we check it here? Is it possible move it before XLogReadRecord? To avoid unnecessary reading.